design-documents/graph-ql/coverage/customer-orders.md

```diff
@@ -0,0 +1,196 @@
+
+### Payment Method Schema
+
+To provide more customization for different payment solutions, the payment method will be represented by own type instead of simple string:
+
+```graphql
+#  payment method used to pay for the order
+type PaymentMethod {
+    name: String! # payment method name for e.g Braintree, Authorize etc
+    type: String! # payment method type used to pay for the order for e.g Credit Card, PayPal etc.
+    additional_data: [KeyValue] # additional data per payment method type
+}
+```
+
+### Pricing Schema
+
+As entities like order, invoice, credit memo might have complex pricing type:
+
+```graphql
+interface OrderPricingInterface {
+    subtotal: Float! # order subtotal excluding, shipping, discounts and tax
+    discounts: [Discount] # all the discounts applied to the order
+    tax: Float! # tax applied to the order
+    grand_total: Float! # final order total including shipping and taxes
+}
+​
+type OrderPricing implements OrderPricingInterface {
+​   shipping_handling: Float! # shipping and handling for the order
+}
+```
+
+## Invoice Type Schema
+
+The invoice entity will have the similar to the order schema:
+
+```graphql
+type Invoice {
+    number: String! # user friendly identifier for the invoice
+    pricing: OrderPricingInterface! # invoice pricing details
+    items: [InvoiceItem]! # invoiced product details
+}
+
+type InvoiceItem implements SalesProductInterface{
+    invoiced: Float! # number of invoiced items
+    children: [InvoiceChildItem]
+}
+
+type InvoiceChildItem implements SalesProductInterface{
+    invoiced: Float! # number of invoiced items
+}
+
+type InvoicePricing implements OrderPricingInterface {
+  
+}
+```
+
+## Refund Type Schema
+
+The credit memo entity will have the similar to the order and invoice schema:
+
+```graphql
+type CreditMemo {
+    number: String!
+    items: [CreditMemoItem]! # items refunded
+    pricing: OrderPricingInterface! # refund pricing details
+}
+
+type CreditMemoItem implements SalesProductInterface{
+    refunded: Float! # number of refunded items
+    children: [CreditMemoChildItem]
+}
+type CreditMemoChildItem implements SalesProductInterface{
+    refunded: Float! # number of refunded items
+}
+
+type CreditMemoPricing implements OrderPricingInterface {
+
+}
+```
+
+## Shipment Type Schema
+
+```graphql
+type OrderShipment {
+    shipping_method: String! # shipping method for the order
+    shipping_address: CustomerAddress! # shipping address for the order
+    tracking_link: String # tracking link for the order
+    shipped_items: [ShipmentItem] # items included in the shipment
+}
+ 
+type ShipmentItem implements SalesProductInterface{
+    shipped: Float! #number of shipped items
+    children: [ShipmentChildItem]
+}
+
+type ShipmentChildItem implements SalesProductInterface{
+    shipped: Float! # umber of shipped items
+}
+```
+
+## Additional Types
+
+The `KeyValue` type will provide a possibility to use key-value pairs:
+
+```graphql
+type KeyValue {
```

Comment:
It would be better to introduce specific types.

Reply:
The idea is to have generic key => value type because we have a lot of custom data which doesn't have a strict structure, for example, payment details, each payment integration has own set of additional data and we should provide a possibility to customize it by 3rd party integrations.